doc/guide/di.md

```diff
@@ -0,0 +1,146 @@
+# Dependency Injection
+
+There are many different ways to use AngularDart to provide and utilize dependency injection throughout your application, components, and services. This guide is meant to serve as a **development tool** and to help push **best practices** from the developers of the AngularDart framework.
+
+* [Providers](#providers)
+    * [DO use the "named" providers](#do-use-the-named-providers)
+    * [DO use factories for configuration](#do-use-factories-for-configuration)
+    * [DO use `const` providers](#do-use-const-providers)
+* [Injectors](#injectors)
+    * [AVOID using `ReflectiveInjector`](#avoid-using-reflectiveinjector)
+
+## Providers
+
+### DO use the "named" providers
+
+While not formally deprecated, `provide(...)` and `Provider(...)` require certain combinations of named optional parameters, and allow combinations that otherwise will not compile or perform as intended. For example:
+
+**BAD**:
+
+```dart
+// This is valid with the older syntax. What does it mean?
+const Provider(
+  Foo,
+  useClass: CachedFoo,
+  useExisting: OtherFoo,
+  useFactory: createCachedFoo,
+  useValue: null,
+);
+```
+
+Instead, use the "named" providers (`ClassProvider`, `ExistingProvider`, `FactoryProvider`, and `ValueProvider`) in order to make your providers more readable, in some cases more terse, and use the definition of the class to help guide the configuration.
+
+**GOOD**:
+
+```dart
+const ClassProvider(Foo, CachedFoo);
```

Comment:
Should be

```dart
const ClassProvider(Foo, useClass: CachedFoo);
```

as `useClass` is an named, optional parameter. I believe this is to support `const ClassProvider(Foo)` when you actually want to use `Foo`. 

The question is should this syntax exist, as is it not just a verbose way of writing only `Foo` in a list of providers?

Reply:
Oh, whoops, is because of the `multi` API, got it (thought it was positional optional and not named optional - I'll make an edit).

> as is it not just a verbose way of writing only `Foo` in a list of providers?

We want to move away from "lists" of providers, so yeah, we'd want to make sure that all providers are concretely of the type `Provider`.

Reply:
While I generally prefer named, optional parameters, this is a case where using a positional optional parameter would make the API more consistent with the other variants. Obviously this isn't possible at the moment, due to the named `multi` parameter. Will this eventually be removed now that we have `MultiToken`, but exists currently to ease migration to these new provider types?

Reply:
Yup. I'll make sure note `MultiToken` is preferred in the guide,